### PR TITLE
[IMP] New behavior for compute to allow read-write.

### PR DESCRIPTION
**Pull request by kujiu**
On stored computed fields with `readonly=False`, do not compute the field if a value is passed by write or create method.

Instead of
```python
    fiscal_position = fields.Many2one('account.fiscal.position')

    @api.onchange('partner_id')
    def onchange_partner_fiscal_position(self):
        self.fiscal_position = self.partner_id.property_account_position
```
we define
```python
    fiscal_position = fields.Many2one('account.fiscal.position',
                                      compute='_compute_fiscal_position',
                                      store=True, readonly=False)

    @api.depends('partner_id.property_account_position')
    def _compute_fiscal_position(self):
        for record in self:
            record.fiscal_position = record.partner_id.property_account_position
```
So the onchange is defined as a computed field that can be modified by the user.

Behavior
    - an onchange triggered on 'partner_id' automatically invalidates 'fiscal_position' and recomputes it
    - write({'partner_id': pid}) automatically invalidates 'fiscal_position' and recomputes it
    - write({'partner_id': pid, 'fiscal_position': fp}) does not recompute 'fiscal_position'
    - create({'partner_id': pid}) automatically computes 'fiscal_position'
    - create({'partner_id': pid, 'fiscal_position': fp}) does not recompute 'fiscal_position'

The behavior covers:
    - onchange methods: potentially a majority of onchange methods can be expressed as compute methods
    - create() and write() behave as if they had executed the onchange methods

### CODE REVIEW

File: doc/reference/orm.rst [hunk 1]
```diff
@@ -429,6 +429,30 @@ it uses the values of other *fields*, it should specify those fields using
             record.discount_value = discount
             record.total = record.value - discount
 
+* stored computed fields are read-only by default. If the ``readonly``
+  parameter is set to ``False``, the field is computed if a dependency
+  changes and if a new value is not given.::
```
Comment (odony) [inline doc/reference/orm.rst]: `versionadded`?

**Comment by odony** [comment]
One thing that seems important with this feature, **before merging it**, is to actually use it in a real business case - as a deeper smoke test. Perhaps we could select one App that has a lot of `@api.onchange` methods and see how we can replace them with read-write computed fields.
What do you think?

**Comment by pedrobaeza** [comment]
I see a potential problems with this, which is performance: each create/write will trigger all the related computed methods, which can be a performance drainer if you don't want to use that "extra" feature. Possible use case: bulk import of thousand/millions of records where you control all the fields you want to write and you don't want the extra time it takes for the compute. There should be a way to inhibit this behavior, for example through a context key.

On the other side, I love this feature and opt-in mechanism with `readonly=False` option for using the feature, but being backwards compatible with "old" computed fields that are `readonly=True` by default is a great decision.

We will still need onchange methods for returning warnings or fields domains, so it's not yet the end of onchanges, but virtually!

**Comment by gdgellatly** [comment]
my main concerns are two fold

1. Nested onchanges - should be obvious what I mean - and I think this is probably covered.
2. Multiple onchanges -
There is a relatively common idiom that goes something like When a main field is entered change these 3 fields, A, B and C.  Lets say partner changes warehouse, sales_team and fiscal position.  But if field D (lets say channel) is changed then sales team must be set to False or some other value and fields E and F also change.

Now as I understand it, you could just use onchange IF you controlled the entire codebase.  However if you are building an extension module I fear (perhaps wrongly) you might get boxed in a corner here by a decision to use compute upstream.

**Comment by pedrobaeza** [comment]
> 1. Nested onchanges - should be obvious what I mean - and I think this is probably covered.

This is covered through normal dependency chain through `@api.depends`

> 2. Multiple onchanges -

No problem, but you have to change your approach a bit, as now you isolate on each compute in one side dependencies, and in the other updated fields. An onchange mixes both. But this isn't anything new. Just convert properly onchange methods to compute ones.

**Pull request closed without merging**